### ISSUE TITLE
helmfile template --selector not works in some cases

### ISSUE THREAD

**Issue by katsew** [roboll/helmfile#1818]
Hi,

I use helmfile template to the same release and filter by labels to output different manifests within the same release.
However, `--selector` not filter the target release on master.

Here is the sample command and helmfile.yaml to reproduce this.

```sh
helmfile --selector index=1 template --output-dir "./out" 

# I also tried
helmfile --selector index=1 template --include-needs=[false|true] --output-dir "./out" 
```

```yaml
releases:
  - name: foo
    chart: incubator/raw
    namespace: default
    labels:
      app: test
      component: raw
      index: '1'
    values:
      - resources:
          - value: 1

  - name: foo
    chart: incubator/raw
    namespace: default
    labels:
      app: test
      component: raw
      index: '2'
    values:
      - resources:
          - value: 2
```

expected output is this:

```
---
# Source: raw/templates/resources.yaml
metadata:
  labels:
    app: raw
    chart: raw-0.2.5
    heritage: Helm
    release: foo
value: 1 # same as --selector index=1
```

however, actual output is this:

```
---
# Source: raw/templates/resources.yaml
metadata:
  labels:
    app: raw
    chart: raw-0.2.5
    heritage: Helm
    release: foo
value: 2 # expected 1
```

It seems that selector not works and render all releases in the same namespace, so the last one win.
I looked into the tree and It seems that #1772 is the one this change introduced.

UPDATE:
I also try appending argument `--include-needs=true|false`, since #1772  mentioned about that, I still not work.

**Comment by katsew** [comment]
I dig into this issue and I see that `withDAG` func cannot detect duplicated release name,
since it convert release to id, so cannot filter the same release name even if the selector applied.
How can I fix this... 🤔

**Comment by mumoshu** [comment]
@katsew Hey! Thanks for trying our main branch. That's super helpful.

Regarding the one you found, I thought we had some duplication-checking mechanism that resulted in Helmfile failing-fast when you have duplicated release that shares the same kubecontext, namespace, and name.

https://github.com/roboll/helmfile/blob/5a3bd7d64919c0a8fd921cfcbd5f01202525d970/pkg/app/app.go#L1013

So, your example should have failed with a duplicate-release error, but it didn't. That's the first issue for me. Probably that's a regression introduced by #1772?

Second, why do you want to use the selector for de-duplication? Depending on your requirement/goal, I can probably guide you to re-design your configuration or form a feature-request to helmfile.

**Comment by mumoshu** [comment]
Well, I gave it a try with the latest released version of helmfile and it did work as you've expected anyway! I definitely need to recall my memory, and see how it changed the behaviour.

```
$ helmfile1387 -f helmfile.1818.yaml -l index=1 template
Templating release=foo, chart=incubator/raw
---
# Source: raw/templates/resources.yaml
metadata:
  labels:
    app: raw
    chart: raw-0.2.5
    heritage: Helm
    release: foo
value: 1
```

**Comment by mumoshu** [comment]
Okay, so it turned out that the duplication-checking logic was skipped in helmfile-template since #1772, as the logic was in `processFilteredReleases` which is not called anymore in the code-path since #1772.

> I dig into this issue and I see that PlanReleases func cannot detect duplicated release name,
since it convert release to id, so cannot filter the same release name even if the selector applied.

Now I came back to this sentence which makes sense.

The point is that on this line:

https://github.com/roboll/helmfile/blob/5a3bd7d64919c0a8fd921cfcbd5f01202525d970/pkg/app/app.go#L1714

The releases being planned here expected to have no duplicate releases, because the DAG production logic is not aware of duplicate release names/IDs.

What we're basically doing with `PlanReleases` here is that "Do produce a DAG of all the releases contained in `st`, but omitting all the releases not included in `selectedReleases`".

I admit it's a bit confusing, but theoretically speaking the fix would be to make this "all the releases contained in `st`" deduplicated, which means we need to deduplicate `st.Releases` before running `st.PlanReleases`.

**Comment by katsew** [comment]
@mumoshu Hi. Thanks for the reply.

First, let me explain my use case which is why I hit into this issue.
I have a chart which relies on full-stack web application framework.
It has many roles to change its behavior, like API, WebUI, TaskQueue, CronJob, and need to deploy independently.
Even if we deploy each role independently, we need the same configs to ship with, so I cannot split my chart right now.
Here is my pseudo releases which has flags for each roles to achieve my use case.

```
releases:
  - name: app
    chart: my-chart
    namespace: default
    labels:
      app: my-chart
      component: api
    values:
      - api.enabled: true

  - name: app
    chart: my-chart
    namespace: default
    labels:
      app: my-chart
      component: webui
    values:
      - ui.enabled: true

  - name: app
    chart: my-chart
    namespace: default
    labels:
      app: my-chart
      component: task-queue
    values:
      - queue.enabled: true
 ...
```

Then, I use the selector to release each role like this:

```
helmfile  -l app=my-chart,component=api template
```

Before using helmfile, I use helm with a complex shell script to fit in my case, and I try replacing that script with helmfile right now.
I've tried several ways to get helmfile work for my case like, change kubeContext for each components, separate helmfile.yaml, templating helmfile with condition, etc., but these are a bit tricky...
I thought that using selector is one way to fit in my case without introducing tricky things into helmfile.

**Comment by katsew** [comment]
> The point is that on this line:
> 
> https://github.com/roboll/helmfile/blob/5a3bd7d64919c0a8fd921cfcbd5f01202525d970/pkg/app/app.go#L1714
> 
> The releases being planned here expected to have no duplicate releases, because the DAG production logic is not aware of duplicate release names/IDs.
> 
> What we're basically doing with `PlanReleases` here is that "Do produce a DAG of all the releases contained in `st`, but omitting all the releases not included in `selectedReleases`".
> 
> I admit it's a bit confusing, but theoretically speaking the fix would be to make this "all the releases contained in `st`" deduplicated, which means we need to deduplicate `st.Releases` before running `st.PlanReleases`.
I think I have better understanding of what `PlanReleases` does, thanks.
Let me find the way to deduplicate `st.Releases`.

**Comment by katsew** [comment]
> Okay, so it turned out that the duplication-checking logic was skipped in helmfile-template since #1772, as the logic was in `processFilteredReleases` which is not called anymore in the code-path since #1772.
> 
> This is the logic that is not called anymore in helmfile-template since #1712

I don't understand why `processFilteredReleases` not called since #1772  in the first place?
I thinks to add this func may solve this issue.

**Comment by mumoshu** [comment]
@katsew I have not yet rechecked my theory by running helmfile locally, but I thought your patch would break when you have `needs` i.e. when the DAG calculation actually come into play.

As far as I can recall right now, `SetFilter` filters out unselected releases at the time of loading your helmfile.yaml. That does deduplicate `st.Releases` by selectors- true. On the other hand, though, unselected releases are now missing in the `st.Releases` that results in `PlanReleases` failing due to needed releases are missing in `st.Releases`.

To recap, `PlanReleases` needs a complete list of releases. The list needs to be deduplicated in terms of kubecontext/namespace/name. But it also needs to contain all the releases that are expected to be parts of the DAG(this is especially important when you use `needs`). Filtering releases by the selector at the load time breaks the last point.

**Comment by katsew** [comment]
hmm... that's true, I'll find another way, thanks for the recap.

**Comment by mumoshu** [comment]
> On the other hand, though, unselected releases are now missing in the st.Releases that results in PlanReleases failing due to needed releases are missing in st.Releases.

@katsew I've reread the code carefully and this turned out to be not particularly right.

`PlanReleases` does fail when needed releases are missing. But #1772 should have allowed that by forcing the user to provide a valid combination of `--selector`, `--include-needs`, `--skip-needs` to let `PlanReleases` receive necessary and sufficient releases and options to build the DAG. Actually, that was the primary purpose of #1772.

So, back to your finding:

> I dig into this issue and I see that PlanReleases func cannot detect duplicated release name,
since it convert release to id, so cannot filter the same release name even if the selector applied.
How can I fix this... 🤔

As it turned out we can safely pass a subset of releases to the dag generation logic, we can let the deduplicated list of releases here:

https://github.com/roboll/helmfile/blob/f28ad5af2330621ac85e1ddf05c526c4147f2340/pkg/app/app.go#L1742-L1754

by changing `st.Releases = allReleases` to `st.Releases = selectedReleasesWithNeeds `.

Note that it must be `selectedReleasesWithNeeds` rather than `selectedReleases`, `helm-template` won't run on releases included by `--include-needs`.

All these findings are reflected into #1822. I'd appreciate it if you could review it. Thanks in advance for your support!

**Comment by katsew** [comment]
Hi, @mumoshu .
I verified your fix with my case described in this issue, and it seems that the fix does not work.
Would you mind reopening this issue?

**Comment by mumoshu** [comment]
@katsew Hey! Thanks for confirming.

Not sure how I slipped it while testing, but I've reproduced it myself, too. Running some debugging sessions to see what's going on now...

**Comment by mumoshu** [comment]
@katsew We have two `PlanReleases`, one directly, another via `withDAG`. Apparently, I've missed treating the first case.

**Comment by mumoshu** [comment]
@katsew It seems like we unintentionally opened up a Pandora's box, probably by combining #1772 and #1814 😅 

I've written my best understanding of all the issues and how we are able to fix em in #1823.

It would be great if you could review/test it. Thanks

**Comment by katsew** [comment]
@mumoshu 

Hi, thanks for your great job!
LGTM for my case, but I found a case which may not be intentional.

here is a releases for that case:

```
releases:
  - name: foo
    chart: incubator/raw
    namespace: default
    labels:
      app: test
      component: raw
      index: '1'
    values:
      - resources:
          - value: 1
    needs:
      - default/foo-test

  - name: foo
    chart: incubator/raw
    namespace: default
    labels:
      app: test
      component: raw
      index: '2'
    values:
      - resources:
          - value: 2

  - name: foo-test
    chart: incubator/raw
    namespace: default
    labels:
      app: test
      component: raw
      index: '2'
    values:
      - resources:
          - kind: PriorityClass

```

when I run like `helmfile --selector index=2 template --output-dir "./out"`,
I expected that only output foo with label index=2, but it also produce foo-test...

Maybe this is because `PlanReleases` knows only id produced by ReleaseToId?

**Comment by katsew** [comment]
@mumoshu 

Sorry! The case show you above is correct.
Both foo-test and second foo have label index=2, and it is expected, meaning your fix is LGTM!

**Comment by mumoshu** [comment]
@katsew Thank you so much for testing! That's super helpful.

I'll go merging the mentioned PR 🤞
